src/client/auth/aws.rs

```diff
@@ -0,0 +1,414 @@
+             {token}\
+             x-mongodb-gs2-cb-flag:n\n\
+             x-mongodb-server-nonce:{nonce}\n\n\
+             {signed_headers}\n\
+             {hashed_body}\
+             ",
+            host = host,
+            date = date_str,
+            token = token,
+            nonce = nonce,
+            signed_headers = signed_headers,
+            hashed_body = hashed_body,
+        );
+
+        let hashed_request = hex::encode(Sha256::digest(request.as_bytes()));
+
+        let small_date = date.format("%Y%m%d").to_string();
+
+        let region = if host == "sts.amazonaws.com" {
+            "us-east-1"
+        } else {
+            let parts: Vec<_> = host.split('.').collect();
+            parts.get(1).copied().unwrap_or("us-east-1")
+        };
+
+        #[rustfmt::skip]
+        let string_to_sign = format!(
+            "\
+             AWS4-HMAC-SHA256\n\
+             {full_date}\n\
+             {small_date}/{region}/sts/aws4_request\n\
+             {hashed_request}\
+            ",
+            full_date = date_str,
+            small_date = small_date,
+            region = region,
+            hashed_request = hashed_request,
+        );
+
+        let first_hmac_key = format!("AWS4{}", self.secret_key);
+        let k_date = hmac(first_hmac_key, &small_date)?;
+        let k_region = hmac(k_date, region)?;
+        let k_service = hmac(k_region, "sts")?;
+        let k_signing = hmac(k_service, "aws4_request")?;
+
+        let signature_bytes = hmac(k_signing, string_to_sign)?;
+        let signature = hex::encode(signature_bytes);
+
+        #[rustfmt::skip]
+        let auth_header = format!(
+            "\
+             AWS4-HMAC-SHA256 \
+             Credential={access_key}/{small_date}/{region}/sts/aws4_request, \
+             SignedHeaders={signed_headers}, \
+             Signature={signature}\
+            ",
+            access_key = self.access_key,
+            small_date = small_date,
+            region = region,
+            signed_headers = signed_headers,
+            signature = signature
+        );
+
+        Ok(auth_header)
+    }
+}
+
+fn hmac(key: impl AsRef<[u8]>, input: impl AsRef<[u8]>) -> Result<impl AsRef<[u8]>> {
+    let mut hmac = Hmac::<Sha256>::new_varkey(key.as_ref())?;
+    hmac.input(input.as_ref());
+
+    let output = hmac.result();
+    Ok(output.code())
+}
+
+/// The response from the server to the `saslStart` command in a MONGODB-AWS authentication attempt.
+struct ServerFirst {
+    conversation_id: Bson,
+    server_nonce: Vec<u8>,
+    sts_host: String,
+    done: bool,
+}
+
+impl ServerFirst {
+    /// Parses the response of the `saslStart` command.
+    fn parse(response: Document) -> Result<Self> {
+        let SaslResponse {
+            conversation_id,
+            payload,
+            done,
+        } = SaslResponse::parse("MONGODB-AWS", response)?;
+
+        let mut payload_document = Document::from_reader(&mut payload.as_slice())?;
```

Comment:
I think this could be a good opportunity to leverage `serde` to avoid all the type and presence checking boilerplate here. We'd have to map any errors to use `Error::invalid_authentication_response` as part of that.

Reply:
Yeah looking at the scram code it does seem to avoid `serde`. I honestly don't have any idea why; maybe I just wasn't that familiar with it back then or something. Whatever the reason, current me thinks that `serde` would be the way to go for this kind of stuff.

Reply:
So I think updating this to use serde is the only remaining thing for me before LGTM.

Reply:
The server uses IDL since I trust a machine generated parser more then a user-written one.